### PR TITLE
chore: initial refactor for limit orders

### LINKED ISSUE #7908: Initial refactoring for limit orders
### Overview

**TL;DR - move swapper specific stuff to higher order components**

### References and additional details

Originally spiked in https://github.com/shapeshift/web/issues/7905

The following changes are likely required to prepare the `TradeInput` component and its children for reuse with limit orders:

- Compose the existing `TradeInputHeader`, `TradeInputBody`, `ConfirmSummary` components without `Acknowledgement` wrappers and `CollapsibleQuoteList`, and pass all feature-specific callbacks and data in as props.
- General cleanup - Move functions like `getQuoteErrorTranslation` to a shared utilities file.
- Within `TradeInputHeader`, `TradeInputBody`, `ConfirmSummary` components, replace `useAppSelector` and `useAppDispatch` calls with props where state or logic is not shared.
- Likely have to make the "cog" menu button contents (on the top-right) a prop too, since limit orders will have different items.

### Acceptance Criteria

Task list above is completed.

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Refactors trade input components to allow us easy build-out of limit orders.

Likely plenty more can be done to streamline things for limit orders component-wise, but this PR is intended to be an 80:20 of value:effort so many of the known improvements are intentionally left out where the effort of changing these are one of:

- Simpler to do at the time of building limit orders
- Not strictly required

## Issue (if applicable)

closes #7908 

## Risk

> High Risk PRs Require 2 approvals

High risk. The changes here are theoretically straightforwards but considering the impact of getting things wrong and breaking the ability to make trades, we should review this under a "high risk" lens.

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

All trades for all protocols, namely UI transitions and interactions.

Additionally, arbitrum bridge which is not directly affected but may be indirectly affected UI-wise.

## Testing

Swaps, bridges, quotes, should be EXACTLY the same as prod.

Check trades, namely UI transitions and interactions around:
- trade input
- quote list
- confirming a trade
- settings "cog" menu (custom slippage etc)
- account selection, trading into specific account no.
- manual receive address
- mobile UI layout

Additionally test the arbitrum bridge which is not directly affected but may be indirectly affected UI-wise.

### Engineering

Rebase was finicky, so please double check regression of https://github.com/shapeshift/web/pull/7896/files

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

## Screenshots (if applicable)

A trade

https://jam.dev/c/6f6d30de-65eb-49e3-8c8d-20b621c4e283

Mobile layout unchanged

image

### CODE REVIEW

File: src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx [hunk 1]
```diff
@@ -247,92 +258,52 @@ export const TradeInput = ({ isCompact, tradeInputRef }: TradeInputProps) => {
...
+              sellAsset={sellAsset}
+              sideComponent={CollapsibleQuoteList}
+              tradeInputRef={tradeInputRef}
+              walletReceiveAddress={walletReceiveAddress}
+              handleSwitchAssets={handleSwitchAssets}
+              onSubmit={handleTradeQuoteConfirm}
+              setBuyAsset={setBuyAsset}
+              setBuyAssetAccountId={setBuyAssetAccountId}
+              setSellAsset={setSellAsset}
+              setSellAssetAccountId={setSellAssetAccountId}
```
Comment (woodenfurniture) [inline src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx:302]: image

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Code-wise this is looking good.

I did a regression test with this branch (left) against `develop` (right) and found a few small but important things:

❌ **Going to the Claim tab and back again leads to a an empty Trade/Bridge route** 

Screenshot 2024-10-11 at 16 19 43

✅ **End-to-end trade flow (including approval)**

Screenshot 2024-10-11 at 16 23 59

❌ **Padding broken**

We are missing the padding (48px) that was wrapping the Trade component

Screenshot 2024-10-11 at 16 28 43

❓ **Infinite quote loading is back, and a weird flicker in the quote list - note this happened once and I couldn't repro again, so I"m not going to pin it to this branch**

Screenshot 2024-10-11 at 16 12 49

https://github.com/user-attachments/assets/3041c502-87d9-48e4-8a10-d36400912f94

**Review by 0xApotheosis — APPROVED** [review]
Retested - issues from the previous review have been resolved, looking good.

File: src/components/MultiHopTrade/components/SharedTradeInput/SharedTradeInput.tsx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            rightContent={headerRightContent}
+            onChangeTab={onChangeTab}
+          />
+          <SharedTradeInputBody
+            activeQuote={activeQuote}
+            buyAmountAfterFeesCryptoPrecision={buyAmountAfterFeesCryptoPrecision}
+            buyAmountAfterFeesUserCurrency={buyAmountAfterFeesUserCurrency}
+            buyAsset={buyAsset}
+            initialBuyAssetAccountId={initialBuyAssetAccountId}
+            initialSellAssetAccountId={initialSellAssetAccountId}
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/SharedTradeInput/SharedTradeInput.tsx:94]: suggestion: while these are called *initial* in destructuring of `useAccountId()` return value here:

https://github.com/shapeshift/web/blob/8272547a1fb78a64d436c0489dda52e32defdda3/src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx#L82-L87

There is not much of `initial`:

https://github.com/shapeshift/web/blob/253ed2c6d25c3a3c36b9613d22ba4d9f995d4d1e/src/components/MultiHopTrade/hooks/useAccountIds.tsx#L15-L17

I got quite confused while working in this domain, as this is really the *selected* AccountIds, do we want to use this terminology instead as part of the refactor?

File: src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx [hunk 3]
```diff
@@ -28,61 +28,83 @@ import { isKeplrHDWallet } from 'lib/utils'
...
+  const translate = useTranslate()
   const mixpanel = getMixPanel()
   const history = useHistory()
   const { showErrorToast } = useErrorHandler()
+  const { manualReceiveAddress, walletReceiveAddress } = useReceiveAddress({
+    fetchUnchainedAddress: Boolean(wallet && isLedger(wallet)),
+  })
+  const {
+    sellAssetAccountId: initialSellAssetAccountId,
+    buyAssetAccountId: initialBuyAssetAccountId,
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx:84]: See comment above re: do we want to stop destructure-renaming this, or perhaps use `selected` terminology instead?

File: src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx [hunk 4]
```diff
@@ -91,35 +113,13 @@ export const TradeInput = ({ isCompact, tradeInputRef }: TradeInputProps) => {
...
   )
 
-  const shouldShowTradeQuoteOrAwaitInput = useAppSelector(selectShouldShowTradeQuoteOrAwaitInput)
-  const isTradeQuoteRequestAborted = useAppSelector(selectIsTradeQuoteRequestAborted)
-  const hasUserEnteredAmount = useAppSelector(selectHasUserEnteredAmount)
-  const activeQuote = useAppSelector(selectActiveQuote)
-
-  const isSnapshotApiQueriesPending = useAppSelector(selectIsSnapshotApiQueriesPending)
-  const votingPower = useAppSelector(state => selectVotingPower(state, votingPowerParams))
+  const isKeplr = useMemo(() => !!wallet && isKeplrHDWallet(wallet), [wallet])
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx:116]: Note for a follow-up: we can probably remove this, historical reasons were Osmosis swaps were quite borky in Keplr (or not implemented, can't recall exactly) iirc, but with swapper v2 and custom receive addies, this should now work (and will even better once we bring EVM support to Keplr)

Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx:116]: https://github.com/shapeshift/web/pull/2971 for historical ref

**Review by gomesalexandre — COMMENTED** [review]
First conceptual pass, one smolish comment re: `useAccountIds()` destructured vals renaming but nothing obviously standing out here, runtime pass to follow!

**Review by gomesalexandre — APPROVED** [review]
Tested against *prod* (right, note that is *not* develop because of current ws issues in develop endpoints) and with *app* env, couldn't find any regression in swapper either presentational or functional!

- Trade input is happy (desktop)

image

- Trade input is happy (mobile)

image

- Claims are happy (mobile)

image

- Claims are happy (desktop)

image

- Quotes list (desktop)

image

- Quotes list (mobile)

image

- Cog settings wheel is happy

https://jam.dev/c/21445392-954b-4017-b643-39305a9afd30

- Trade confirm/execution is happy

https://jam.dev/c/7b57952c-cd00-404b-b134-2d8eecfc667d

- Multi-account and manual receive address swaps are happy

https://jam.dev/c/847a8b92-3d7d-4f1b-9b30-9c3b9a9508b9

**Pull request merged**